### PR TITLE
Return a 404 if the requested basin doesn't exist

### PR DESCRIPTION
**Pull request by ee-usgs**
Before making a pull request
----------------------------

- [x] Make sure all tests run
- [x] Update the changelog appropriately

Fixed:  Getting a 500 when making a /basin call to NLDI (WDFN-334)
-----------
https://internal.cida.usgs.gov/jira/browse/WDFN-334

After making a pull request
---------------------------
- [x] If appropriate, put the link to the PR in the JIRA ticket
- [x] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java [hunk 1]
```diff
@@ -93,11 +95,15 @@ protected void streamFeatures(HttpServletResponse response,
...
-		streamResults(transformer, BaseDao.BASIN, parameterMap);
+		if (comid != null) {
+			Map<String, Object> parameterMap = new HashMap<>();
+			parameterMap.put(Parameters.COMID, NumberUtils.parseNumber(comid, Integer.class));
+			BasinTransformer transformer = new BasinTransformer(response);
+			addContentHeader(response);
+			streamResults(transformer, BaseDao.BASIN, parameterMap);
+		} else {
+			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Basin was not found");
+		}
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:106]: Allowing a null comid through throws an IllegalArgumentException at line 100, which isn't specific enough to know what to do with later.  Thus, throwing a ResponseStatusException when it is null instead.

File: src/main/java/gov/usgs/owi/nldi/controllers/GlobalDefaultExceptionHandler.java [hunk 2]
```diff
@@ -38,8 +39,12 @@ public static String handleError(Exception ex, HttpServletResponse response) {
 				//This exception's message contains implementation details after the new line, so only take up to that.
 				return ex.getLocalizedMessage().substring(0, ex.getLocalizedMessage().indexOf("\n"));
 			} else {
-				return ex.getLocalizedMessage().replaceAll("([a-zA-Z]+\\.)+","");
+				return ex.getLocalizedMessage().replaceAll("([a-zA-Z]+\\.)+", "");
 			}
+		} else if (ex instanceof ResponseStatusException) {
+			ResponseStatusException rse = (ResponseStatusException)ex;
+			response.setStatus(rse.getStatus().value());
+			return rse.getLocalizedMessage();
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/owi/nldi/controllers/GlobalDefaultExceptionHandler.java:47]: Following the pattern of the other exceptions, only the status (404) is returned in the response.  It might be nice to include more info in the response, such as 'The basin could not be found'.  A user would assume that, but it would allow them to be confident of what is happening.

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java (hunk 1 above)
Comment (dsteinich) [reply src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:106]: I would leave this method alone. This is called from the LinkedDataController. The next method there (getFlowlines) has a better pattern for checking the comid.

File: src/main/java/gov/usgs/owi/nldi/controllers/GlobalDefaultExceptionHandler.java (hunk 2 above)
Comment (dsteinich) [reply src/main/java/gov/usgs/owi/nldi/controllers/GlobalDefaultExceptionHandler.java:47]: This is class only meant for unhandled exceptions. An expected (reasonable) not found condition should be handled in the code where it occurs.

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java (hunk 1 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:106]: OK, I see that.  I didn't notice example in there before.

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java [hunk 3]
```diff
@@ -124,7 +124,20 @@ protected boolean isLegacy(String legacy, String navigationMode) {
...
+	 * @param featureSource
+	 * @param featureID
+	 * @return May return null if the COM ID is not found.
+	 */
 	protected String getComid(String featureSource, String featureID) {
+
+		if (null == featureSource || null == featureID) {
+			throw new IllegalArgumentException("A featureSource and featureID are required");
+		}
+
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:140]: To distinguish null parameters (a server error) from a NOT FOUND comId, we need to explicitly test for null parameters.  In normal use, the incoming parameters are never null, so if either are null it is really a server error,

File: src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java [hunk 4]
```diff
@@ -78,35 +80,47 @@ public void getComidTest() {
...
 		verify(logService).logRequestComplete(any(BigInteger.class), any(int.class));
+		//Mock lookupDao 1st response of null means the comid is not found, thus a 404
 		assertEquals(HttpStatus.NOT_FOUND.value(), response.getStatus());
 
 		controller.getFlowlines(request, response, null, null, null, null, null, null);
 		verify(logService, times(2)).logRequest(any(HttpServletRequest.class));
 		verify(logService, times(2)).logRequestComplete(any(BigInteger.class), any(int.class));
-		//this is a INTERNAL_SERVER_ERROR because of NPEs that shouldn't happen in real life.
+		//Mock lookupDao 2nd response doesn't actually exist, thus causes a 500 when we try to get flowlines
 		assertEquals(HttpStatus.INTERNAL_SERVER_ERROR.value(), response.getStatus());
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java:107]: This second part of this test (and others like it) is really confusing because we are artificially creating the INTERNAL_SERVER_ERROR as a side-effect of mocking the lookupDao to return a non-existing id.  Really it just tests how the system handles the mocks since the actual two requests are identical.

Best world:  It would be nice to clean up these tests by adding real lookup data for these tests to hit.

File: src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java [hunk 5]
```diff
@@ -132,12 +146,32 @@ public void getCharacteristicDataTest() throws IOException {
 
 	@Test
 	public void getBasinTest() throws Exception {
+		when(lookupDao.getComid(anyString(), anyMap())).thenReturn(goodFeature());
+		doNothing().when(streamingDao).stream(anyString(), anyMap(), any());
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java:150]: The streamingDao will fail because the 'goodFeature' doesn't actually exist.  Here I just force a mock streamingDao to not complain.  This test is now the happy path - the previous test of null input values doesn't really test a situation we would see (and its tested directly by getComId).

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java (hunk 3 above)
Comment (dsteinich) [reply src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:140]: As a style note, all four places that this method are called force a value for both parameters. This could be viewed as unnecessary duplicate code.

File: src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java (hunk 4 above)
Comment (dsteinich) [reply src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java:111]: This entire class is unit testing the controller. The second response is showing that we have captured the unanticipated error, converting it to an INTERNAL_SERVER_ERROR (500) and thus not leaking implementation details to the outside world.

File: src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java (hunk 5 above)
Comment (dsteinich) [reply src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java:154]: As noted above, this class is pure unit tests of the controller itself. Full fledged integration tests against the database are in LinkedDataControllerOtherIT. A not found test there may be appropriate.

File: src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java (hunk 3 above)
Comment (ee-usgs) [reply src/main/java/gov/usgs/owi/nldi/controllers/BaseController.java:140]: Agreed, a null cannot come through from the end points.  But this makes for a clear distinction between invalid arguments and Not Found, as well as makes it easier to detect it in the calling methods.

File: src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java (hunk 5 above)
Comment (ee-usgs) [reply src/test/java/gov/usgs/owi/nldi/controllers/LinkedDataControllerTest.java:154]: Good thought - I'll add that on Friday.
